cl/corpus_importer/management/commands/columbia_merge.py

```diff
@@ -0,0 +1,587 @@
+                        find_just_name(op.author_str).lower(),
+                        find_just_name(author_str).lower(),
+                    )
+                    if s.ratio() >= 0.6:
+                        # columbia names are better
+                        op.author_str = author_str
+                    else:
+                        raise AuthorException(f"Authors don't match")
+                elif any(s.isupper() for s in op.author_str.split(",")):
+                    # Some names are uppercase, update with processed names
+                    op.author_str = author_str
+
+        converted_text = convert_columbia_html(
+            file_opinion["opinion"], columbia_pos
+        )
+        op.html_columbia = str(converted_text)
+        op.save()
+
+
+def map_and_merge_opinions(
+    cluster_id: int,
+    columbia_opinions: list[dict],
+) -> None:
+    """Map and merge opinion data
+
+    :param cluster_id: Cluster id
+    :param columbia_opinions: list of columbia opinions from file
+    :return: None
+    """
+
+    cl_cleaned_opinions = get_cl_opinion_content(cluster_id)
+
+    if len(columbia_opinions) == len(cl_cleaned_opinions):
+        # We need that both list to be cleaned, so we can have a more accurate match
+        matches = match_opinion_lists(
+            [
+                clean_opinion_content(op["opinion"], is_harvard=False)
+                for op in columbia_opinions
+            ],
+            [op.get("opinion") for op in cl_cleaned_opinions],
+        )
+        if len(matches) == len(columbia_opinions):
+            update_matching_opinions(
+                matches, cl_cleaned_opinions, columbia_opinions
+            )
+        else:
+            raise OpinionMatchingException("Failed to match opinions")
+
+    elif len(columbia_opinions) > len(cl_cleaned_opinions) == 1:
+        for op in columbia_opinions:
+            opinion_type = op.get("type")
+            file = op.get("file")
+            if not opinion_type:
+                raise OpinionTypeException(
+                    f"Opinion type unknown: {op.get('type')} found in: {file}"
+                )
+            author = op.get("byline")
+
+            converted_text = convert_columbia_html(op["opinion"], op["order"])
+
+            Opinion.objects.create(
+                html_columbia=converted_text,
+                cluster_id=cluster_id,
+                type=opinion_type,
+                author_str=titlecase(find_just_name(author.strip(":")))
+                if author
+                else "",
+            )
+
+            logger.info(f"Opinion created for cluster: {cluster_id}")
+
+    else:
+        # Skip creating new opinions due to differences between data, this may happen
+        # because some opinions were incorrectly combined when imported with the old
+        # columbia importer, or we have combined opinions
+        logger.info(
+            msg=f"Skip merging mismatched opinions on cluster: {cluster_id}"
+        )
```

Comment:
this should never be called right?  How would we get here?

Reply:
This can happen due to combined opinions, for example, here we have 4 opinions: https://www.courtlistener.com/opinion/1440110/texas-department-of-mental-health-mental-retardation-v-rodriguez/ and the xml file has only 3 opinions (texas/court_opinions/documents/5e79ca463f540b2d.xml)

That is not considered in the conditions:` len(cl_cleaned_opinions) > len(columbia_opinions)` that is why it falls in the else

We could exclude the combined opinion as a last resort and try to match the opinions again.

Reply:
yes - but 1, I dont see it throwing an error and stopping the code.  And 2 wouldnt it throw an error in mismatched opinions earlier.